File: tkpdflight/src/main/java/com/tokopedia/flight/orderlist/domain/model/FlightOrderPassengerViewModelMapper.java
package com.tokopedia.flight.orderlist.domain.model;

import com.tokopedia.flight.booking.view.viewmodel.FlightBookingAmenityViewModel;
import com.tokopedia.flight.orderlist.data.cloud.entity.PassengerAmentityEntity;
import com.tokopedia.flight.orderlist.data.cloud.entity.PassengerEntity;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

/**
 * Created by alvarisi on 12/12/17.
 */

public class FlightOrderPassengerViewModelMapper {
    @Inject
    public FlightOrderPassengerViewModelMapper() {
    }

    public FlightOrderPassengerViewModel transform(PassengerEntity passengerEntity) {
        FlightOrderPassengerViewModel viewModel = null;
        if (passengerEntity != null) {
            viewModel = new FlightOrderPassengerViewModel();
            viewModel.setPassengerFirstName(passengerEntity.getFirstName());
            viewModel.setPassengerLastName(passengerEntity.getLastName());
            viewModel.setPassengerBirthdate(passengerEntity.getDob());
            viewModel.setPassengerTitleId(passengerEntity.getTitle());
            viewModel.setType(passengerEntity.getType());
            viewModel.setAmenities(transformAmenities(passengerEntity.getAmenities()));
        }
        return viewModel;
    }

    private List<FlightBookingAmenityViewModel> transformAmenities(List<PassengerAmentityEntity> amenities) {
        List<FlightBookingAmenityViewModel> viewModels = new ArrayList<>();

        for (PassengerAmentityEntity amentityEntity : amenities) {
            FlightBookingAmenityViewModel amenityViewModel = new FlightBookingAmenityViewModel();
            amenityViewModel.setId(String.valueOf(amentityEntity.getSequence()));
            amenityViewModel.setPrice(amentityEntity.getPrice());
            amenityViewModel.setPriceNumeric(amentityEntity.getPriceNumeric());
            amenityViewModel.setTitle(amentityEntity.getDetail());
            amenityViewModel.setDepartureId(amentityEntity.getDepartureAirportId());
            amenityViewModel.setArrivalId(amentityEntity.getArrivalAirportId());
            amenityViewModel.setAmenityType(amentityEntity.getAmenityType());
            viewModels.add(amenityViewModel);
        }
        return viewModels;
    }

    public List<FlightOrderPassengerViewModel> transform(List<PassengerEntity> entities) {
        List<FlightOrderPassengerViewModel> viewModels = new ArrayList<>();
        FlightOrderPassengerViewModel viewModel;
        for (PassengerEntity entity : entities) {
            viewModel = transform(entity);
            if (viewModel != null) {
                viewModels.add(viewModel);
            }
        }
        return viewModels;
    }
}

File: features/travel/flight_orderlist/src/main/java/com/tokopedia/flight/orderlist/domain/model/mapper/FlightOrderPassengerViewModelMapper.kt
package com.tokopedia.flight.orderlist.domain.model.mapper

import com.tokopedia.flight.orderlist.constant.FlightCancellationStatus
import com.tokopedia.flight.orderlist.data.cloud.entity.CancellationEntity
import com.tokopedia.flight.orderlist.data.cloud.entity.PassengerAmentityEntity
import com.tokopedia.flight.orderlist.data.cloud.entity.PassengerEntity
import com.tokopedia.flight.orderlist.domain.model.FlightOrderPassengerViewModel
import com.tokopedia.flight.orderlist.view.viewmodel.FlightOrderAmenityViewModel
import java.util.*
import javax.inject.Inject

/**
 * Created by alvarisi on 12/12/17.
 */

class FlightOrderPassengerViewModelMapper @Inject
constructor() {

    fun transformAmenities(amenities: List<PassengerAmentityEntity>): List<FlightOrderAmenityViewModel> {
        return amenities.map {
            val amenityViewModel = FlightOrderAmenityViewModel()
            amenityViewModel.id = it.sequence.toString()
            amenityViewModel.price = it.price
            amenityViewModel.priceNumeric = it.priceNumeric
            amenityViewModel.title = it.detail
            amenityViewModel.departureId = it.departureAirportId
            amenityViewModel.arrivalId = it.arrivalAirportId
            amenityViewModel.amenityType = it.amenityType
            return@map amenityViewModel
        }
    }

    fun transform(entities: List<PassengerEntity>): List<FlightOrderPassengerViewModel> {
        return entities.map {
            return@map FlightOrderPassengerViewModel(
                    it.type,
                    if (it.cancelStatus.isNotEmpty()) it.cancelStatus[0].status else 0,
                    if (it.cancelStatus.size > 1) it.cancelStatus[1].status else 0,
                    it.title,
                    it.firstName,
                    it.lastName,
                    transformAmenities(it.amenities),
                    if (it.cancelStatus.isNotEmpty()) it.cancelStatus[0].statusStr else "",
                    if (it.cancelStatus.size > 1) it.cancelStatus[1].statusStr else "")
        }
    }

    fun getCancelledPassengerCount(cancellations: List<CancellationEntity>): Int {
        val cancelledPassengers = ArrayList<String>()

        for (item in cancellations) {
            for (attribute in item.details) {
                if (!cancelledPassengers.contains(attribute.passengerId + attribute.journeyId) && (item.status == FlightCancellationStatus.PENDING ||
                                item.status == FlightCancellationStatus.REFUNDED ||
                                item.status == FlightCancellationStatus.REQUESTED)) {
                    cancelledPassengers.add(attribute.passengerId + attribute.journeyId)
                }
            }
        }

        return cancelledPassengers.size
    }
}
